src/Orchard/Tasks/Locking/Services/DistributedLockService.cs

```diff
@@ -0,0 +1,153 @@
+using System;
+using System.Data;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Autofac;
+using Orchard.Data;
+using Orchard.Environment;
+using Orchard.Exceptions;
+using Orchard.Logging;
+using Orchard.Services;
+using Orchard.Tasks.Locking.Records;
+
+namespace Orchard.Tasks.Locking.Services {
+
+    public class DistributedLockService : IDistributedLockService {
+        private readonly IMachineNameProvider _machineNameProvider;
+        private readonly ILifetimeScope _lifetimeScope;
+        private readonly IClock _clock;
+        private readonly object _semaphore = new object();
+
+        public DistributedLockService(IMachineNameProvider machineNameProvider, ILifetimeScope lifetimeScope, IClock clock) {
+            _machineNameProvider = machineNameProvider;
+            _lifetimeScope = lifetimeScope;
+            _clock = clock;
+            Logger = NullLogger.Instance;
+        }
+
+        public ILogger Logger { get; set; }
+
+        public bool TryAcquireLock(string name, TimeSpan maxLifetime, TimeSpan timeout, out IDistributedLock @lock) {
+            lock (_semaphore) {
+                @lock = default(IDistributedLock);
+
+                try {
+                    var waitedTime = TimeSpan.Zero;
+                    var waitTime = TimeSpan.FromMilliseconds(timeout.TotalMilliseconds / 10);
+                    bool acquired;
+
+                    while (!(acquired = TryAcquireLockRecord(name, maxLifetime, out @lock)) && waitedTime < timeout) {
+                        Task.Delay(timeout).ContinueWith(t => {
+                            waitedTime += waitTime;
+                        }).Wait();
+                    }
+
+                    if (acquired) {
+                        Logger.Debug("Successfully acquired a lock named {0}.", name);
```

Comment:
I try to always surround arbitrary strings with single quotes to avoid ambiguity in the output message if the string should contain spaces, periods etc.